### ISSUE TITLE
Collection of issues from Force.com toolkit example

### ISSUE THREAD

**Issue by mungojam** [hvanbakel/CsprojToVs2017#165]
I've now run the conversion of the SalesForce example (from shanselman blog) using the [unconditional branch](https://github.com/andrew-boyarshin/CsprojToVs2017/tree/unconditional). There are a few things left in the project file that I thought I had sorted, but also some that might be easy to fix.

You can see all the converted project files in [my fork](https://github.com/mungojam/Force.com-Toolkit-for-NET/blob/converter-v2/src)

Just going to list these here to get opinions on whether they are fixable in the current work or should be issues for future work:

[ChatterToolkitForNet.csproj](https://github.com/mungojam/Force.com-Toolkit-for-NET/blob/converter-v2/src/ChatterToolkitForNET/ChatterToolkitForNET.csproj)

Could we remove this?:
```xml
    <MinimumVisualStudioVersion>11.0</MinimumVisualStudioVersion>
```

I thought we can remove project GUIDs?:
```xml
    <ProjectGuid>{011B1F31-45AA-4CD1-BB97-F4A508408365}</ProjectGuid>
```

We could provide an option to automatically convert to netstandard, or to remove portable project settings if the consumer chooses to upgrade to netstandard
```xml
    <TargetFrameworkProfile>Profile111</TargetFrameworkProfile>
    <ProjectTypeGuids>{786C830F-07A1-408B-BD7F-6EE04809D6DB};{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}</ProjectTypeGuids>
```

This feels like something auto-generated, so can we detect and remove it?
```xml
    <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\</SolutionDir>
```

Pretty sure this is default nowadays
```xml
    <RestorePackages>true</RestorePackages>
```

Not sure on these, but I would have thought we can generally ditch them. Maybe we need a safe mode and an eager mode!
```xml
    <ProductVersion>12.0.0</ProductVersion>
    <SchemaVersion>2.0</SchemaVersion>
    <OutputPath>bin\$(Configuration)\</OutputPath>
```

These are now the ones that I thought I had set to be removed if they were the same because msbuild and VS only need one of them if they all match. I had worked out which one, but maybe hadn't put it into a test.

```xml
    <RootNamespace>Salesforce.Chatter</RootNamespace>
    <AssemblyName>Salesforce.Chatter</AssemblyName>
    <AssemblyTitle>Salesforce.Chatter</AssemblyTitle>
```

I think before we didn't output these if they were blank, but maybe it was only if they were missing completely from assemblyinfo
```xml
    <Description />
```

Not sure what this one is about as the AssemblyInfo contains `[assembly: AssemblyConfiguration("")]`. I think it's rarely used and we could probably detect that default and then remove this and the setting in AssemblyInfo
```xml
    <GenerateAssemblyConfigurationAttribute>false</GenerateAssemblyConfigurationAttribute>
```

We could remove these since they haven't been changed from the default. They will get added back in if the user ever wants to change the versions themselves.
```xml
    <AssemblyVersion>1.0.0.0</AssemblyVersion>
    <FileVersion>1.0.0.0</FileVersion>
```

These feel like a default setting, can we remove them? Is it that the new default is portable pdbs? Maybe the eager option could get rid of them.
```xml
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
    <DebugType>full</DebugType>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
    <DebugType>pdbonly</DebugType>
  </PropertyGroup>
```

same comment about portable libraries as above
```xml
  <Import Project="$(MSBuildExtensionsPath32)\Microsoft\Portable\$(TargetFrameworkVersion)\Microsoft.Portable.CSharp.targets" />
```

Lots of empty `ItemGroup`s
```xml
  <ItemGroup />
  <ItemGroup />
  <ItemGroup>
    <ProjectReference Include="..\CommonLibrariesForNET\CommonLibrariesForNET.csproj" />
  </ItemGroup>
  <ItemGroup />
</Project>
```

**Comment by andrew-boyarshin** [comment]
Collection of what I consider to be safe to remove/fix:
```xml
<MinimumVisualStudioVersion>...</MinimumVisualStudioVersion>
<SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\</SolutionDir>
<RestorePackages>true</RestorePackages>
<SchemaVersion>2.0</SchemaVersion>
<Description />
<XVersion>1.0...</XVersion>
<!-- The following is indeed a regression on unconditional branch to be fixed in next PRs -->
<ItemGroup />
<!-- AssemblyConfiguration attribute in AssemblyInfo and related MSBuild property -->
```
The following is unsafe to remove since we still support PCLs.
```xml
<Import Project="$(MSBuildExtensionsPath32)\Microsoft\Portable\$(TargetFrameworkVersion)\Microsoft.Portable.CSharp.targets" />
```
The following is unsafe to remove since user might have custom build targets that depend on executables/dlls, not adapted to Portable PDBs format. If we remove them, we break such scenarios. Portable PDBs are still pretty new and support is flaky.
```xml
<DebugType>full</DebugType>
<DebugType>pdbonly</DebugType>
```
The following is an actual user data, no way we can remove that.
```xml
<ProductVersion>12.0.0</ProductVersion>
```
The following was default in legacy CPS indeed, and as with Portable PDBs custom build targets or some CI configuration might depend on this setting. No way it can be removed automatically.
```xml
<OutputPath>bin\$(Configuration)\</OutputPath>
```
The following is the same as with Portable PDBs - too dangerous to remove.
```xml
<ProjectGuid>{011B1F31-45AA-4CD1-BB97-F4A508408365}</ProjectGuid>
```
The following is unsafe to remove since it defines project system features: ASP.NET Razor, WinForms, WPF, generic XAML support.
```xml
<ProjectTypeGuids>...</ProjectTypeGuids>
```

**Comment by mungojam** [comment]
It's looking healthier now: [ChatterToolkitForNET.csproj](https://github.com/mungojam/Force.com-Toolkit-for-NET/blob/with-pr-167/src/ChatterToolkitForNET/ChatterToolkitForNET.csproj)

> And automatic conversion to .NET Standard is mostly infeasible, since there are no mappings to .NET Standard for the majority of PCL profiles.

I thought that .netstandard2.0 covers the vast majority of them though, so we could offer to update it as long as no unsupported ones are present. Or we could at least do it if the user manually chooses to update to .net standard as part of the conversion.

**Comment by andrew-boyarshin** [comment]
Heads-up on remaining work after #170 merge.

`ProjectGuid` removal can be scheduled for 2.1, as of #170 we now store the value from solution, so if it matches we can remove it from project file.

PCL profiles to .NET Standard conversion can be 2.1 scheduled too. There is about 12 supported PCL profiles (of approximately 80), but they are the most common.

`ProjectTypeGuids` unification (all to upper-case) can be scheduled for 2.1.

So-called _eager_ mode (I'd call it tooling modernization), like removal of `OutputPath` and `DebugType` can be implemented in 2.x.

`<Description />` is still an issue in 2.0 after #170.

`<XVersion>1(.0)</XVersion>` removal can be scheduled for 2.1.

```xml
<Import Project="$(MSBuildExtensionsPath32)\Microsoft\Portable\$(TargetFrameworkVersion)\Microsoft.Portable.CSharp.targets" />```
The one above can be made autoremoved in 2.x if no PCL profiles are left in TargetFrameworks.

These are all remaining issues mentioned in your issue description, that should (in my opinion) actually be fixed in the future.

**Comment by mungojam** [comment]
I like that plan!

> <Description /> is still an issue in 2.0 after #170.

I thought I had fixed that in #167

**Comment by andrew-boyarshin** [comment]
@mungojam oh, my bad. When I was writing my message I only checked the PR description and it wasn't obvious from that.

**Comment by hvanbakel** [comment]
Can we spin the individual workitems out so we actually have something to work towards? Maybe a list like in #154?

**Comment by hvanbakel** [comment]
Closing this, will track work in individual items.

**Comment by mungojam** [comment]
Thanks @hvanbakel, things have got busier for me now that I'm back from hols
